js/codap-handler.js

```diff
@@ -214,10 +218,41 @@ export default class CodapHandler {
       })
   }
 
+  // determine the run number of the last case in the collection.
+  retrieveRunNumber (runNumberCallback) {
+    const runsCollection = `dataContext[${DATA_SET_NAME}].collection[RunSummary]`
+    // determine the number of cases
+    codapInterface.sendRequest({
+      action: 'get',
+      resource: `${runsCollection}.caseCount`
+    }, responseCallback)
+    .then((response) => {
+      if (response.success) {
+        // request the last case
+        const caseCount = response.values
+        if (!caseCount) return
+        codapInterface.sendRequest({
+          action: 'get',
+          resource: `${runsCollection}.caseByIndex[${caseCount - 1}]`
+        }, responseCallback)
+        .then((response) => {
+          // retrieve the run number from the last case
+          if (response.success) {
+            const theCase = response.values['case']
+            const runNumber = theCase.values[config.others.run.codapDef.name]
+            if (runNumber) {
+              runNumberCallback(runNumber)
+            }
+          }
+        }, responseCallback)
+      }
+    })
+  }
```

Comment:
That's a perfectly reasonable suggestion and that was my first thought as well. There are a couple of issues here, however, the combination of which led me to choose this approach.

1. Saving of the game state to LARA/CODAP document is a completely independent operation from sending the data to CODAP via the plugin API. Given their independence, it is conceivable that they might get out of sync at some point. The page could be reloaded between the CODAP data export and the LARA save, for instance. If they were ever to get out of sync, the value returned by the CODAP plugin API is the correct one, as any other value would result in duplicate run numbers or skipped run numbers in the exported data, which would immediately appear to the user as a bug.

2. Despite the above (because I don't think I had thought it through completely at that point), I initially started down the path of adding the run number to the saved document state. This turned out to be less straightforward than I expected because the only values being saved at this point are game state values (`challengeIdx` and `stepIdx`); the names of the relevant functions all refer to gameState; and those functions are called at times determined by the needs of the game; the `gameState` accessor function hard-codes the value of `game` to `true`. In other words, the existing state-saving infrastructure is tied pretty tightly to game mode. The `runNumber` value we need to save is only (currently) an issue in simulation mode. (It's used under the hood in game mode, but never exposed to the user.) Therefore, to save `runNumber` would require re-factoring the gameState code to explicitly support simulation mode as well.

The re-factoring suggested in item (2) may be desirable in its own right, but given that it would require more QA due to the re-factoring involved, and it would still be subject to potential synchronization bugs as a result of (1), I decided to take the path of using the alternate retrieval path. This is analogous to what other plugins like `SensorInteractive` and `SageModeler` are already doing to keep themselves synced with saved documents.

Reply:
Great, thanks for the clarification! It makes sense. :)